Ml exec contribution (#2472)
Codecademy/docs: content/python/concepts/built-in-functions/terms/exec/exec.md

--
Title: 'exec()'
Description: 'Executes a code object or string containing python code.'
Subjects:
  - 'Computer Science'
  - 'Data Science'
Tags:
  - 'Functions'
  - 'Methods'
  - 'Strings'
CatalogContent:
  - 'learn-python-3'
  - 'paths/computer-science'
  - 'paths/data-science'
---

The **`exec()`** function executes a code object or string containing python code.

## Syntax

```pseudo
exec(object, globals=None, locals=None)
```

* `object` : String or code object
* `globals`: (optional) A dictionary containing global variables. If not specified, defaults to `None`. 
* `locals`: (optional) A dictionary containing global variables. If not specified, defaults to `None`.

The `exec()` function returns `None`.

## Example 1

This example uses `exec()` to parse and execute Python code contained in the string `code`:

```py
code = 'print("Hello, Codecademy!")'

exec(code)

> ```

Reviewer: I'd like an output block after the example. It should look like this:

````
This example results in the following output:

```shell
Hello, Codecademy!
```
````

## Example 2

This example uses `exec()` to execute python code from a file `code.txt`, which contains Python commands:

```py
#code.txt
import datetime

current_time = datetime.datetime.now()

print(current_time)
```

The content of the file `code.txt` is read until the end of the file (EOF) into a string. The content (Python commands) is executed by `exec()`.

```py
with open('code.txt','r') as file:
  code = file.read()

exec(code)
```

> ## Example 3

Reviewer: I'd like you to swap this codebyte example with example 4 so this comes at the end. Then change the header to read `## Codebyte Example"

* `globals`: (optional) A dictionary containing global variables. If not specified, defaults to `None`. 
* `locals`: (optional) A dictionary containing global variables. If not specified, defaults to `None`.

The `exec()` function returns `None`.

## Example 1

This example uses `exec()` to parse and execute Python code contained in the string `code`:

```py
code = 'print("Hello, Codecademy!")'

exec(code)
```

This example results in the following output:

```shell
Hello, Codecademy!
```

## Example 2

This example uses `exec()` to execute python code from a file `code.txt`, which contains Python commands:

```py
#code.txt
import datetime

current_time = datetime.datetime.now()

print(current_time)
```

The content of the file `code.txt` is read until the end of the file (EOF) into a string. The content (Python commands) is executed by `exec()`.

```py
with open('code.txt','r') as file:
  code = file.read()

exec(code)
```

This example will produce output like the following, but with the current date:

```shell
2023-08-16 14:17:06.092145
```

## Example 3

This example uses `exec()` with `globals` and `locals`:

```py
code = input("Enter your command: ")

> #In this example a user could inject malicious code like "import os; os.system('cat /etc/passwd')"

Reviewer 2: ```suggestion
# In this example a user could inject malicious code like "import os; os.system('cat /etc/passwd')"
```

exec(code)

code = input("Enter another command: ")

> #The globlas dictonary allows only the print function from the buitlins.

Reviewer 2: ```suggestion
# The globals dictionary allows only the print function from the builtins.
```

exec(code,{"__builtins__": {"print": print}},{})

def f1():
    print('Hello, Codecademy!')

def f2():
    print('Hello, world!')

> #locals restrict the usage of f1 function with exec:

Reviewer 2: ```suggestion
# locals restrict the usage of f1 function with exec:
```

exec("f1()",{"__builtins__": {}}, {"f1": f1})

> exec("f2()",{"__builtins__": {}}, {"f1": f1}) #This will throw an error

Reviewer 2: ```suggestion
exec("f2()",{"__builtins__": {}}, {"f1": f1}) # This will throw an error
```

--
Title: 'exec()'

> Description: 'Executes code object or string containing python code.'

Reviewer: ```suggestion
Description: 'Executes a code object or string containing python code.'
```

Subjects:
  - 'Computer Science'
  - 'Data Science'
Tags:
  - 'Functions'
  - 'Methods'
  - 'Strings'
CatalogContent:
  - 'learn-python-3'
  - 'paths/computer-science'
  - 'paths/data-science'
---

> Returns ``None``.

Reviewer: ```suggestion
The **`exec()`** function executes a code object or string containing python code.
```

## Syntax

> ```py

Reviewer: ````suggestion
```pseudo
````

exec(object, globals=None, locals=None)
```

> * ```object```: String or code object

Reviewer: ```suggestion
* `object` : String or code object
```
Inline code is only a single backtick.

> * ```globals```: (optional) A dictionary containing global variables. If not specified, defaults to ```None`. 

Reviewer: ```suggestion
* `globals`: (optional) A dictionary containing global variables. If not specified, defaults to `None`. 
```

> * ```locals```: (optional) A dictionary containing global variables. If not specified, defaults to ```None```.

Reviewer: ```suggestion
* `locals`: (optional) A dictionary containing global variables. If not specified, defaults to `None`.
```

> 

Reviewer: ```suggestion

The `exec()` function returns `None`.

```
Moved this here from the intro sentence above.

## Example 1

> Use `exec()` to parse and execute Python code contained in the string `code`:

Reviewer: ```suggestion
This example uses `exec()` to parse and execute Python code contained in the string `code`:
```

> ```codebyte/python

Reviewer: ````suggestion
```py
````
We generally only have one codebyte at the end of the entry.

Instead, add a ` ```shell ` block with the output after the example along with an introductory sentence. Like this:

````
This example results in the following output:

```shell
Hello, Codecademy!
```
````
```

code = 'print("Hello, Codecademy!")'

exec(code)
```

## Example 2

> Use `exec()` to execute python code from a file `code.txt`, which contains Python commands:

Reviewer: ```suggestion
This example uses `exec()` to execute python code from a file `code.txt`, which contains Python commands:
```

```py
#code.txt
import datetime

current_time = datetime.datetime.now()

print(current_time)
```

The content of the file `code.txt` is read until the end of the file (EOF) into a string. The content (Python commands) is executed by `exec()`.

> ```codebyte/python

Reviewer: ````suggestion
```py
````
See my above comment about one codebyte per entry. Also this wouldn't work for the user as there's be no `code.txt` file to access on the server.

with open('code.txt','r') as file:
  code = file.read()

exec(code)
```

## Example 3

> Use `exec()` to execute a code object:

Reviewer: ```suggestion
This example uses `exec()` to execute a code object:
```

```codebyte/python
code = 'import datetime\nprint("Hello, Codecademy!")\ncurrent_time = datetime.datetime.now()\nprint(current_time)'

code_object = compile(code,'<string>','exec')

exec(code_object)
```

## Example 4

> Use `exec()` with `globals` and `locals`:

Reviewer: ```suggestion
This example uses `exec()` with `globals` and `locals`:
```

> ```codebyte/python

Reviewer: ````suggestion
```py
````
